### PR TITLE
`const` modifier on type parameters

### LINKED ISSUE #30680: Suggestion: Const contexts for generic type inference
## Search Terms

const contexts; literals; narrowing; generics; 

## Suggestion

Allow const contexts for generic type parameter inference.

Often I've written and seen generic functions where the type parameter inference is intended to be as narrow as possible.  Accomplishing this involves several "tricks" with generic constraints.  It is more reminiscent of alchemy than I would like, and places quite a burden on the function signature and anyone with the misfortune of having to read and understand it:

```ts
// 🧙 what sorcery is this?? 🧙‍ 
type Narrowable = string | number | boolean | symbol | object | undefined | void | null | {};
declare function foo<N extends Narrowable, T extends { [k: string]: N | T | [] }>(x: T): T; 

foo({ a: 1, b: "c", d: ["e", 2, true, { f: "g" }] });
// T inferred as { a: 1; b: "c"; d: ["e", 2, true, { f: "g"; }]; }
```

Const contexts (#29510) are exactly the knob we want to turn here, and the "`as const`" syntax 
is succinct, understandable, and non-mind-bending.  Unfortunately, the only way to use this in
generic functions is from the caller's side, which is hard to guarantee:

```ts
declare function bar<T extends object>(x: T): T; 
bar({ a: 1, b: "c", d: ["e", 2, true, { f: "g" }] } as const); // burden on function caller
// T inferred as { 
//   readonly a: 1; readonly b: "c"; d: readonly ["e", 2, true, { readonly f: "g"; }]; 
// }

bar({ a: 1, b: "c", d: ["e", 2, true, { f: "g" }] }); // oops!!
// T inferred as { 
//   a: number; b: string; d: (string | number | boolean | { f: string; })[]; 
// } 😢

```

The suggestion here is to get the best of both worlds by allowing a const context to be specified in the generic type parameter declaration:

```ts
declare function baz<T extends const object>(x: T): T; // 🤔
declare function baz<T extends object as const>(x: T): T; // 🤷
declare function baz<const T extends object>(x: T): T; // 😵
declare function baz<const T const extends readonly object as const>(x: T): T; // 🧠💥🤪

// also possible, see #46937:
declare function baz<T extends object>(x: const T): T; 



baz({ a: 1, b: "c", d: ["e", 2, true, { f: "g" }] });
// T inferred as { 
//   readonly a: 1; readonly b: "c"; d: readonly ["e", 2, true, { readonly f: "g"; }]; 
// } ❤🎉
```

## Related issues

#29510: const contexts

#10676: generics infer literals "T extends string | number | boolean"
#27179: generics infer tuples "T extends U[] | [U]"
#13347: probably can't make generics infer readonly 

#16896: please narrow all object literals as much as possible

## Checklist

My suggestion meets these guidelines:

* [x] This wouldn't be a breaking change in existing TypeScript/JavaScript code
* [x] This wouldn't change the runtime behavior of existing JavaScript code
* [x] This could be implemented without emitting different JS based on the types of the expressions
* [x] This isn't a runtime feature (e.g. library functionality, non-ECMAScript syntax with JavaScript output, etc.)
* [x] This feature would agree with the rest of [TypeScript's Design Goals](https://github.com/Microsoft/TypeScript/wiki/TypeScript-Design-Goals).

EDIT: mentioning slightly different approach from #46937 where modifier goes on the type parameter instead of its constraint.  It's not obvious to me if one has an advantage over the other.

### LINKED ISSUE #41114: Support Const Type Constraint
## Search Terms
Const Type Constraint Literal

## Suggestion
Add new syntax `const T` or a global generic `type Constant<T = any> = T extends const T ? T : never;`

`type ConstString = const string` is same as `type ConstString = Constant<string>`.

* If `T` extends `string | number | boolean | symbol | bigint | null | undefined`, requires `T` to be a constant.
* If `T` is `object` or `Array`, requires every member of  `T` to be a constant.
* If `T` is `enum`, nothing to do, because `enum` is same as `const enum` 

The difference with union literal is that type `'foo' | 'bar'` can only be 'foo' or 'bar', but type `const string` can be any constant string.

## Use Cases
* Sometimes we want to constrain a type or a parameter to be literal.
* It can help the TS compiler to recognize constant types better

## Examples
```ts
let foo = 'foo'; // string
const bar= 'bar'; // 'bar'

const notConstant: const string = foo; // error
const trulyConstant: const string = bar; // success

type Baz = { baz: string } ;
const baz1: const Baz = { baz: 'baz' }; // success
const baz2: Baz = { baz: 'baz' }; // not a constant
const baz3: Baz  = { baz: 'baz' } as const; // current available, same as baz1

type ConstArray<T> = const Array<T>;
// same as
type AnotherCosntArray<T> = Array<const T>;

function tuple<T extends ConstArray>(...args: T) {
  return args;
}
const t1 = tuple(foo); // error!
const t2 = tuple(bar); // success! typeof t2 should be ['bar']
const t3 = tuple('a', 1, true, ['1'], baz1) // ['a', 1, true, ['1'], { baz: 'baz' }];
const t4 = tuple(baz2) // error!

let variable: const string = 'variable'; // this is possible
variable = 'another variable'; // success
variable = document.title; // error

let reactiveObject: const { foo: string } = { foo: 'foo' }; // success;
reactiveObject.foo = 'bar' // success
```
**In fact, `const` only affects the assignment behavior, and a `const` type is considered to be the same as a non `const` type when read it.**
**A `const type` doesn't mean immutable, just means every primitive value should be literal, so maybe should call it `literal type`**.

## Alternatives

* Only add `extends const` syntax, and add global generic `type Constant<T = any> = T extends const ? T : never;`
* Use `finally` or `static` or `literal`(not exist in current keywords) keyword instead of `const`
* Use `type Literal` instead of `type Constant`.

## Related issues

#30680 Suggestion: Const contexts for generic type inference 

## Checklist

My suggestion meets these guidelines:

* [x] This wouldn't be a breaking change in existing TypeScript/JavaScript code
* [x] This wouldn't change the runtime behavior of existing JavaScript code
* [x] This could be implemented without emitting different JS based on the types of the expressions
* [x] This isn't a runtime feature (e.g. library functionality, non-ECMAScript syntax with JavaScript output, etc.)
* [x] This feature would agree with the rest of [TypeScript's Design Goals](https://github.com/Microsoft/TypeScript/wiki/TypeScript-Design-Goals).

### PR DESCRIPTION
**Pull request by ahejlsberg**
With this PR we implement a new `const` modifier for type parameters. In a function, method, or constructor invocation, when a literal expression in an argument is contextually typed by a `const` type parameter, the literal expression is given the most precise type possible, similar to having applied an `as const` assertion (see #29510). The kinds of literal expressions affected by a `const` contextual type are string, numeric, and boolean literal values, array literals, and object literals.

The `const` modifier is permitted on type parameters of functions, methods, and classes. It is an error to apply the `const` modifier to type parameters of interfaces and type aliases.

Given the function declarations

```ts
declare function foo<const T>(x: T): T;
declare function bar<const T>(obj: [T, T]): T;
```

the invocations

```ts
const x1 = foo('a');
const x2 = foo(['a', ['b', 'c']]);
const x3 = foo({ a: 1, b: "c", d: ["e", 2, true, { f: "g" }] });
const x4 = bar([{ a: 1, b: 'x' }, { a: 2, b: 'y' }]);

```

infer the types

```ts
declare const x1: "a";
declare const x2: readonly ["a", readonly ["b", "c"]];
declare const x3: {
    readonly a: 1;
    readonly b: "c";
    readonly d: readonly ["e", 2, true, { readonly f: "g"; }];
};
declare const x4:
    { readonly a: 1; readonly b: "x"; } |
    { readonly a: 2; readonly b: "y"; };
```

A `const` type parameter has no effect on variables or other kinds of expressions used as function arguments. For example:

```ts
const a = [1, 2, 3];  // number[]
const b = foo(a);  // number[]
```

The example above still requires writing `[1, 2, 3] as const` to infer type `readonly [1, 2, 3]` for `a`.

When a `const` type parameter is constrained to an array type, that array type should include a `readonly` modifier; otherwise, inferences for the type parameter will fail to meet the constraint. For example:

```ts
declare function baz<const T extends readonly unknown[]>(...args: T): T;

const z = baz(1, 'b', { a: 1, b: 'x' });  // readonly [1, "b", { readonly a: 1; readonly b: "x"; }]
```

Without the `readonly` modifier in the constraint, inference defaults to `unknown[]` because an inferred readonly tuple type wouldn't be assignable to `unknown[]`.

Fixes #30680.
Fixes #41114.

### CODE REVIEW

**Comment by ahejlsberg** [comment]
@typescript-bot test this
@typescript-bot user test this inline
@typescript-bot run dt
@typescript-bot perf test faster
@typescript-bot test top100

**Comment by ahejlsberg** [comment]
@typescript-bot perf test this

**Comment by RyanCavanaugh** [comment]
@typescript-bot pack this

**Comment by mon-jai** [comment]
Could this fix #13923 too?

**Comment by Andarist** [comment]
I can't even express how grateful I am for this PR ❤️ thank you, thank you 

I'm toying with it in the provided preview playground and I wonder if there is any way to achieve some extra custom validation on top of the inferred "const type".

This is my simple attempt at this: [TS playground](https://www.staging-typescript.org/play?ts=5.0.0-pr-51865-14#code/C4TwDgpgBAysCGwIGED2A7AZgSwOZQF4oBvAKCgqm3W2G3gBsB+ALigGdgAna3AbnKVOiCO1ZQAShADGqLgBMAPJx7pcAGlgIkaLHgB8pAL6lSoSFABqjbPJGKAKlAgAPJOnnstI3Tlz7CEkEKalp6BjYAclC6RkjnNwgPLwBrCBBUTCgnJig0jKyAOQxCgFcGBngAIwYIRwBtACJhJHZGgF0AtnQIADcILmCObVE2ByaW0Q7jU3kZSq5oTFL0aToMKGlFkQBZeGkAC2o62XRObIT3T28dDD99AApTvzZrBlt7B30ASjZe1Fspi2EF2+yOPQeZEoVBosQiUEi8EimgA9CiOAdUOV5FAqtBUCkhpN2GwodDKPBSUZ1ENoVUqUMTEZvkDtkg9odjpChjFwlFpMioGiMViGDiBlw5JpaJEvOhUMAYVBgAdoMSiSMSUFyRSqTSdRR6SQTNCmSzSEA). I created way more complex examples of this in the past (that were partially working) but they have always fallen apart one way or another, preventing us from shipping those experiments to the production. If I could figure out how to make this simple example work with the new `const` modifier - I could probably implement most of the things I'd like.

I got some promising results with an intersection ([here](https://www.staging-typescript.org/play?ts=5.0.0-pr-51865-14#code/C4TwDgpgBAysCGwIGED2A7AZgSwOZQF4oBvAKCgqm3W2G3gBsB+ALigGdgAna3AbnKVOiCO1ZQAShADGqLgBMAPJx7pcAGlgIkaLHgB8pAL6lSoSFABqjbPJGKAKlAgAPJOnnstI3Tlz7CEkEKalp6BjYAclC6RkjnNwgPLwBrCBBUTCgnJig0jKyAOQxCgFcGBngAIwYIRwBtACJhJHZGgF0AtnQIADcILmCObVE2ByaW0Q7jU3kZSq5oTFL0aToMKGlFkQBZeGkAC2o62XRObIT3T28dDD99AApTvzGoADIrGzskR30ASjYvVQtlMWwgu32Rx6DzIlCoNFiESgkXgkU0AHp0RwDqhyvIoFVoKgUkNJuw2LC4ZR4BSjOohnCqrShiYjH9QdskHtDscYUMYuEotI0VBMdjcQx8QMuHJNLRIl50KhgPCoMADtAyaSRuSglTqbT6fqKEySCY4az2aQgA)) but it fails to deliver granular errors (as it relies on the whole intersection being potentially reduced to `never`).

File: tests/cases/conformance/types/typeParameters/typeParameterLists/typeParameterConstModifiers.ts [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+    P extends readonly [] ? T :
+    P extends readonly [infer A extends keyof T, ...infer Rest] ? GetPath<T[A], Rest> :
+    never;
+
+function set<T, const P extends readonly string[]>(obj: T, path: P, value: GetPath<T, P>) {}
+
+declare let obj: Obj;
+declare let value: "123";
+
+set(obj, ['a', 'b', 'c'], value);
```
Comment (Andarist) [inline tests/cases/conformance/types/typeParameters/typeParameterLists/typeParameterConstModifiers.ts:60]: What I'm wondering about in my comment [here](https://github.com/microsoft/TypeScript/pull/51865#issuecomment-1348513002) could potentially be used to validate the second argument itself. Right now we need to rely on the third one being computed to `never` based on the invalid const path in the second one.

**Comment by ahejlsberg** [comment]
@typescript-bot perf test faster

**Comment by Andarist** [comment]
@ahejlsberg it's surprising that this modifier has an effect on the inferred return types of functions/methods, it doesn't match the behavior of the traditional `as const` assertion:
```ts
const obj = { prop: 1, test: () => 'foo' } as const
//    ^? { readonly prop: 1; readonly test: () => string }

declare function narrow<const T>(a: T): T

const res = narrow({ prop: 1, test: () => 'foo' })
//    ^? { readonly prop: 1; readonly test: () => 'foo' }
```

[TS playground](https://www.staging-typescript.org/play?ts=5.0.0-pr-51865-14#code/MYewdgzgLgBCBGArGBeGBvGAHATiLAXDAIwA0MUAptEQBQCUqAfDAOQBmIIrMAvjAEMIMUJCgAoAPSSYsmAD0A-OPEATSsAA2AnJRjsArmGBQAluBhgdeAO4AeUdBgAVJrQFFn9TysexdwmhWOLa0mLj4RGQU1FB0jCgsHFw8vPRSMnJK4kA)

File: tests/cases/conformance/types/typeParameters/typeParameterLists/typeParameterConstModifiers.ts (hunk 1 above)
Comment (ahejlsberg) [reply tests/cases/conformance/types/typeParameters/typeParameterLists/typeParameterConstModifiers.ts:60]: You could use the `const` type parameter as a type argument in the constraint to implement further validation. For example:

```ts
type StateKeys<T> = T extends { states?: object } ? keyof T['states'] & string : never;

type StateConfig<S extends string = string> = {
    initial?: S,
    states?: Record<S, StateConfig>
}

declare function createMachine<const T extends StateConfig<StateKeys<T>>>(config: T): void
```

File: src/compiler/parser.ts [hunk 2]
```diff
@@ -2814,7 +2814,7 @@ namespace Parser {
             case ParsingContext.ArrayBindingElements:
```
[the remark this answers is not in the document]
Comment (ahejlsberg) [reply src/compiler/parser.ts:2814]: Hmm, yeah, that doesn't look right.

**Comment by ahejlsberg** [comment]
> it's surprising that this modifier has an effect on the inferred return types of functions/methods

It shouldn't. Will fix.

File: tests/cases/conformance/types/typeParameters/typeParameterLists/typeParameterConstModifiers.ts (hunk 1 above)
Comment (Andarist) [reply tests/cases/conformance/types/typeParameters/typeParameterLists/typeParameterConstModifiers.ts:60]: This is a good technique - thank you for sharing it! Building on that I managed to also build a recursive version (as I need it): [TS playground](https://www.staging-typescript.org/play?ts=5.0.0-pr-51865-14#code/C4TwDgpgBAysCGwIGkIgM4B4AqA+KAvFNlBAB5IB2AJulAN5ToJLoD8AXFAPYBGAVhADGwKAF8AUFChsoAazTcAZsQDaAImaII6dQF0oAMibAATgEtKAcylQulCADcIpgNwTQkWCx058REnIqWgYTbXQuPkERcRk1TR9dA3snF3cPcGg4bQBhbkolcytMWxhSCggaOmYLa0ITWqsAGltsbNZy4LoAJWFuU2pMGstm71z8wqt-KF6hfsHbaWHrFulpdog8gqKhsxGmqHhKEFxbU+n6W0tzYHN4ABtOWHclxKfLtahVZChLeUUVG1EnouBstpMSp91j5UBgcBt0N89LhVlCEfDEkjTp9cC9xO4xOlqMJ7vBTNAlABXSgicz5KBCcnaACy8CEAAtLBBMHNKMxiJ1KiEwRMdhtYVg8Ad0XhcLgABS2XmTLjYCQASi4jm45mo6UZEBZbM5DnlH1+lBud3uXHU8HUqK0rC45uk8BdizW11uD1tSm43Adnte4Q9ULW-u4LrEqM+Mc98ek8bE6vcQA). I should clean this up to use a single type parameter or something - but that's beside the point here. 

Thank you ❤️

**Comment by ahejlsberg** [comment]
@typescript-bot perf test faster

**Comment by Thundercraft5** [comment]
In it's current state, functions with type parameters that have `const` modifiers can be [sneaked into interface methods via `typeof` without any errors](https://www.staging-typescript.org/play?ts=5.0.0-pr-51865-14#code/CYUwxgNghgTiAEAzArgOzAFwJYHtXwBIV0BGAHjDwGcN4AVAPgAoA3KCZEALnoEoe6AbgBQwrKgwgYiKGAR0QNAPIAjAFbhaAb2Hx4kmjwwBPAA4gciQsTAkRAX1GVUNfYowAmAe9UbM8AF54LTdDeDYOBACGcPZOeHsRYQB6ZPgAA2cadPgAdxwYAGsqeCgS1AKAW3ZhLNoCOCoSQNDPADoDDCYtAFEuACIABn77Xl09CdT4AD0AflEpnpgYApKACykQABp4FWRaE3N4LBKaLAgIY9REKThgUpKoDLqcw4RcjfwwdggQYGE3tY0GAPC0KNRaIxWHFuHxAjEhKJQJBYAg6kD0F4MSCkuiGopQUEiMCPN14H0hv0EmMJvApnMgA).

I personally feel that it should be allowed on function/constructor and interface methods, as class methods/type parameters are already allowed to use it, especially because it feels intuitive to try to use it whenever there is an opportunity for type inference.

**Comment by RyanCavanaugh** [comment]
I think this being disallowed is just an oversight?
```ts
type $func2 = <const T>(value: T) => T;
```

**Comment by ahejlsberg** [comment]
A nice 1.5-2.5% reduction in check time from the improved caching of contextual types. Only exception is `material-ui` where the savings are overwhelmed by the cost of checking the very complex contextual types for `const` type parameters. I think that is a pretty atypical situation though, so not particularly worried about it.

**Comment by ahejlsberg** [comment]
> I think this being disallowed is just an oversight?

Yes, fixing now.

**Comment by Thundercraft5** [comment]
This is a bit of a QoL thing, but would having `const` be shown in the type parameter representation on hover similar to `in` and `out` be included?

File: src/compiler/parser.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (ahejlsberg) [reply src/compiler/parser.ts:2814]: Now fixed.

**Comment by L13** [comment]
Hi and thank you for the new feature.

I tried to find duplicated values in a tuple for a function param with the new const modifier, but found something unexpected. The first function bar works correct, but throws a type error. If I fix the type error in the second function foo the function doesn't throw an error anymore, but returns a wrong type. The third function baz works correct, but the type requires a spread operator. Is the second function foo a bug?

[TS Playground](https://www.staging-typescript.org/play?ts=5.0.0-pr-51865-14#code/C4TwDgpgBAShC2B7AbhAIgVzAGwJYGMBDYCAZygB4AVLbaCADxIDsATcgJwkNcWexBQMzANbNEAd2YBtALoAaWGQzZgURi3ZDR4qXIB8UALxQAUFCg0c9JhDblpuZgDMIHS+AiKAdL6ev3OFJgWXMoAH4PSHVbeyVSFWBpZgx4ACM3UIsLSLgkVEwcAmIyCiDgRSDE-TCLAC4lfPRaYpJSMrIKqGlfbyrVRSpPWRr6qC4ePgFu3v6u3qs6WQBuU1NQaLRcYKd8NWpaG01Obl5+QWExSRkR40aUZqKiNoPrRQNV9c9LToBBO62O2Yewo0gAjIoAEwjZZQAD0cPGpymgnBUNCX2iVE6ACEAdtgLtgKCIVBSdD9LCEUjJuduuSMRtoNjggBhfFAkHgmHwxETM7TblrJk-YJoDmE4HEj68mkC1GM74s4AAUQlRNBAAZFNqyYpyejKbL+SjurqGWs1tSqAALDiSchue0cRRpDBqQjYUiIKASRAcESkUzOYR7XB8KBpQjuCj4PjBKIQQwACmQnowZAagMlIKGkH0AEooABvWUqjjOqDOf1QbMavNJhoNqAAcgbLagvDIUHEalIxG2zkEwBt0DjzGCHEITjULZNdMuuhuLe8YS4wAwHGYUDT2AzpFWAF81lGOMnpObZAWqXzkXSL-rQtSAALAUgAWkYkD2n4r-tMp7nuaj7Xsad6CpelqmNSaCIGQzAtmoI72hI6h-i6kbukiG5buQEj2swADmUBMuEwahoSEbVj6sbxmozYaHYWjztMi7XAYUCpumma1gS9aeIWJZrhAOHbru+5HlaiL4XwxHrpu24imR1FASBN5yqaKTpG4cjQYibp9jaiAqKwJF2pIUCENuTr-i+b6fgw37AL+zrBogiCqXqZJXupLGCFpGQcLplrUgA6v6gbkVK4bblGABelDjgmDGxMx4EXDo7G3Fxe48XWUqggswyCcWwmiTu3EHqYx4AYQcWeWCPlgbSEGPnpUCvh+X4QD+NkcLV9UPl5jWgdSflmo+QA)

```typescript
type RemoveDuplicates <Tuple extends readonly unknown[], Result extends unknown[]> = 
  Tuple extends [infer Type, ...infer Rest]
  ? Type extends Result[number]
    ? RemoveDuplicates<Rest, Result>
    : RemoveDuplicates<Rest, [...Result, Type]>
  : readonly [...Result, ...Tuple];

type Distinct <Tuple extends readonly unknown[]> = RemoveDuplicates<Tuple, []>;

type TestA = Distinct<[1, 2]>; // readonly [1, 2]

type TestB = Distinct<[1, 1, 2]>; // readonly [1, 2]

type TestC = Distinct<[1]>; // readonly [1]

type TestD = Distinct<[]>; // readonly []

type TestE = Distinct<[0, 0, 1, 1, 2, 2]>; // readonly [0, 1, 2]



// Throws error, but also works
function bar <const Type> (values: Distinct<Type>) { // Error for Distinct<Type>: Type 'Type' does not satisfy the constraint 'readonly unknown[]'.
  return values;
}

bar([0, 1]); // readonly [0, 1]
// @ts-expect-error
bar([0, 1, 1]); // readonly [0, 1]



// Doesn't throw error, but returns wrong type?
function foo <const Type extends readonly unknown[]> (values: Distinct<Type>) {
  return values;
}

// wrong return type ?
foo([0, 1]); // readonly number[]
// but should throw an error
// @ts-expect-error
foo([0, 1, 1]); // readonly number[]



// Works
function baz <const Type extends readonly unknown[]> (values: Distinct<[...Type]>) {
  return values;
}

baz([0, 1]); // readonly [0, 1]
// @ts-expect-error
baz([0, 1, 1]); // readonly [0, 1]
```

File: src/compiler/parser.ts [hunk 3]
```diff
@@ -7742,10 +7743,10 @@ namespace Parser {
      *
      * In such situations, 'permitInvalidConstAsModifier' should be set to true.
```
Comment (juansalvatore) [inline src/compiler/parser.ts:7744]:
```suggestion
     * In such situations, 'permitConstAsModifier' should be set to true.

```

**Comment by Thundercraft5** [comment]
I wanted to clarify on how `const` modifier would narrow a type parameter if it is inferred via a `this` parameter, [currently it doesn't really do anything](https://www.staging-typescript.org/play?ts=5.0.0-pr-51865-14&ssl=11&ssc=10&pln=1&pc=1#code/CYUwxgNghgTiAEAzArgOzAFwJYHtXwxAGcMAeMPE+AFXhAA9DVgj44pg8IBPeAbVTIAtgCMQMALoA+ABQYAFliIAufgDoN1CQEpVfDWq0AoI6EiwEKdNjwFiGAEzlKGGnUYhmrAN4EoAc1USGCxUf3gAX1kFJVVqXRoTM2g4eFDCGEQoMAQAQRgYKG5Sail4byN4KrsSVQxuAAcQHEQajABuIwiTAHoe+ABCGnkEAHIYolH4ClRCRngWgkaxwVFxPgkppXhUHFcoIiIsf1QoEQgEDBx4IRAFHGBR1nHFSYXW+qb4UYFhMUlRmojDMqAASACM8AAvPxwRI1IQSDJtJ1qvA+vAAHoAfiS4BSCHS4iyOXgAHkRAArcCuCpoxGOOrLRYMhydbrAlzwUEOaHwGS+DABVQAIgADCLItoEfYHMjUdUMTi8eZUuZDjR7M5UFRSuVKtUQRgYMhMDgYDIAG5QCDIEBxFEmNENLBNbW617uJgsTUkUhQVDcKTRV5xV4JCZdExG7kAZj5qBAAHdfRgBQBRZTikURaUupryg1VJW4tGcnWuUEAFgTydTGazEoi8AO5XgmfgwVC4VzanzIELaJLQA).

If type parameters inferred via `this` with `const` were to narrow, it would allow type-safe functional programming such as this:
```ts
declare interface Array<T> {
    // Just an example, lots of other possibilities here.
    typeSafeIncludes<const This extends readonly any[], V>(this: This, value: V): value is This[number];
}

[1, 2, 3, 4, 5, 6, 7, 8, 9, 10].typeSafeIncludes(/* some wide value, such as `string | number */)
```

An alternative to that is just to use a wrapper function or class with a `const` type parameter.

**Review by RyanCavanaugh — APPROVED** [review]

**Comment by magic-akari** [comment]
Is this expected?
```TypeScript
class Foo<in out T>{ }
class Bar<const T>{ } // Ok

new class <in out T>{ }
new class <const T>{ }
//        'const' modifier can only appear on a type parameter of a function, method or class
```
[Playground](https://www.staging-typescript.org/play?ts=5.0.0-pr-51865-14#code/MYGwhgzhAEBiD28A8BLAdteBXALtAKgHwDe0AvgFCiQwBCYATksPGhHkaZRWgKYDu0alGioM2DiXI8BQ8COat2BKWSA)

**Comment by ahejlsberg** [comment]
@magic-akari Argh, no that's an oversight. I'll put up a PR to fix.

**Comment by ctjlewis** [comment]
Phenomenal work here.

**Comment by Mlocik97** [comment]
Q: is this somewhat possible with JSDoc?

I would assume something like:

```js
/** @template {readonly<{ [key: string]: string }>} const T */
```

**Comment by Andarist** [comment]
I don't think that modifiers are allowed anyhow in the `@template` today. I've tried a couple of options and they are not interpreted correctly.

You can use this workaround for now though:
```js
/** @type {<const T>(a: T) => T} */
function test(a) {
    return a;
}
```

**Comment by Mlocik97** [comment]
@Andarist doesn't seems to work, it throws error:

```
The type of a function declaration must match the function's signature.
```

**Comment by Andarist** [comment]
@Mlocik97 hard to tell what's going on without seeing the repro case of your problem. I've shared the snippet yesterday after testing it out within the TS codebase (in one of its JS files) and I didn't see any errors caused by this.

**Comment by Mlocik97** [comment]
OK, I tried testing more, and it seems it works fine in CJS, but not ESM. I will try to provide repro soon.

**Comment by Mlocik97** [comment]
hello, I tested it more, and found out why I got the error... TSC correctly handled that type, but VSCode for some reason didn't picked line:

`"typescript.tsdk": "./node_modules/typescript/lib"` when I had workspace opened.

**Comment by SalathielGenese** [comment]
Hi @ahejlsberg ,

I have a hard time (from your initial example) discovering the rule that makes the inferred type of `x4` (a union) vs. that of `x2` (a tuple).

Thanks.

**Comment by Andarist** [comment]
@SalathielGenese notice that both use different functions, `x2` is the `foo`'s result and `x4` is the `bar`'s result.

It's important to check what is the result of the same `bar` call **without** the `const` type parameter:
```ts
declare function bar<T>(obj: [T, T]): T;

const x4 = bar([{ a: 1, b: 'x' }, { a: 2, b: 'y' }]);
// ?^ const x4: { a: number; b: string; }
```

and what is the result of the same thing with a regular `as const` assertion:
```ts
declare function bar<T>(obj: [T, T]): T;

const x4 = bar([{ a: 1, b: 'x' } as const, { a: 2, b: 'y' } as const]);
// ^? const x4: { readonly a: 1; readonly b: "x"; } | { readonly a: 2; readonly b: "y"; }
```

We can see here that it's consistent with the existing behavior of `const` assertions. The union is allowed likely because both arguments have a common supertype or something like that (a call like this fails to typecheck: `bar(['', 10])`)

**Comment by JavaScriptBach** [comment]
This is a great feature. Would you consider allowing it for type aliases as well?

Example use case: https://github.com/Automattic/mongoose/issues/12782

**Comment by RyanCavanaugh** [comment]
@JavaScriptBach huh? There's no straightforward correspondence from this feature to something that would apply to type aliases. `const` type parameters in generic functions change the interpretation of objects in expression-space, but type aliases operate entirely in type-space.

**Comment by JavaScriptBach** [comment]
@RyanCavanaugh My point is that I have generic type aliases that expect to take in const type parameters, e.g. something like

```typescript
type Foo<T> = ...
```

It would be great to be able to type `T` as a const parameter instead of requiring all callers to specify `as const`. I can't speak for implementation, but I think supporting const type parameters for type aliases would be useful for the same reasons that they're useful today for functions, methods, and classes.

**Comment by RyanCavanaugh** [comment]
> It would be great to be able to type T as a const parameter instead of requiring all callers to specify as const

Again, not understanding. You can't write this, it's syntactically illegal:
```ts
type X = Foo<{ x: 3 } as const>;
```

**Pull request merged**